File: clients/web/src/api/endpoints.ts
export const DEFAULT_HTTP_BASE_URL = "http://127.0.0.1:8080";
export const DEFAULT_WS_URL = "ws://127.0.0.1:8081/ws";
export const WEBSOCKET_PROTOCOL_VERSION = 3;

export function normalizeHttpBaseUrl(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error("HTTP endpoint is required.");
  }

  const url = new URL(trimmed);
  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new Error("HTTP endpoint must start with http:// or https://.");
  }

  url.hash = "";
  url.search = "";
  url.pathname = url.pathname.replace(/\/+$/, "");

  const normalized = url.toString();
  return normalized.endsWith("/") ? normalized.slice(0, -1) : normalized;
}

export function buildOpenApiUrl(httpBaseUrl: string): string {
  return `${normalizeHttpBaseUrl(httpBaseUrl)}/docs/openapi.yaml`;
}

export function normalizeWebSocketUrl(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error("WebSocket endpoint is required.");
  }

  const url = new URL(trimmed);
  if (url.protocol !== "ws:" && url.protocol !== "wss:") {
    throw new Error("WebSocket endpoint must start with ws:// or wss://.");
  }

  url.hash = "";
  return url.toString();
}

File: clients/app/lib/core/network/endpoints.dart
const defaultHttpBaseUrl = 'http://127.0.0.1:8080';
const defaultWsUrl = 'ws://127.0.0.1:8081/ws';
const websocketProtocolVersion = 3;

String normalizeHttpBaseUrl(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    throw FormatException('HTTP endpoint is required.');
  }

  final uri = Uri.parse(trimmed);
  if (uri.scheme != 'http' && uri.scheme != 'https') {
    throw FormatException('HTTP endpoint must start with http:// or https://.');
  }

  final normalized = uri.replace(
    query: null,
    fragment: null,
    path: uri.path.replaceAll(RegExp(r'/+$'), ''),
  );
  final result = normalized.toString();
  return result.endsWith('/') ? result.substring(0, result.length - 1) : result;
}

String normalizeWebSocketUrl(String value) {
  final trimmed = value.trim();
  if (trimmed.isEmpty) {
    throw FormatException('WebSocket endpoint is required.');
  }

  final uri = Uri.parse(trimmed);
  if (uri.scheme != 'ws' && uri.scheme != 'wss') {
    throw FormatException(
      'WebSocket endpoint must start with ws:// or wss://.',
    );
  }

  return uri.replace(fragment: null).toString();
}

String buildOpenApiUrl(String httpBaseUrl) {
  return '${normalizeHttpBaseUrl(httpBaseUrl)}/docs/openapi.yaml';
}
